File: thinking-categorically/functor/csv-transform/csv-transform.hs
module Main where

-- CSV Transform example in Haskell using Functor (List functor)
import Data.Char (toUpper)
import Data.List.Split (splitOn)

data Person = Person
  { name     :: String
  , age      :: Int
  , ageGroup :: String
  } deriving (Show)

-- Parse CSV string into list of Person objects
readCSV :: String -> [Person]
readCSV csv =
  case lines csv of
    [] -> []
    header:rows ->
      let keys = splitOn "," header
          parseRow row =
            let values = splitOn "," row
                rowMap = zip keys values
                n = lookup "name" rowMap
                a = lookup "age" rowMap >>= Just . read
            in case (n, a) of
                 (Just nameVal, Just ageVal) ->
                   let group = if ageVal > 30 then "old" else "young"
                   in Just $ Person (map toUpper nameVal) ageVal group
                 _ -> Nothing
      in map (\r -> case parseRow r of
                      Just p -> p
                      Nothing -> error "Malformed row") rows

main :: IO ()
main = do
  let csv = "name,age\nAlice,25\nBob,40"
  let people = readCSV csv
  mapM_ print people

File: thinking-categorically/functor/csv-transform/csv-transform.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Person
{
    public string? Name { get; set; }
    public int Age { get; set; }
    public string? AgeGroup { get; set; }

    public override string ToString() =>
        $"{{ name: '{Name}', age: {Age}, ageGroup: '{AgeGroup}' }}";
}

class Program
{
    static void Main()
    {
        // Parse CSV file into a dictionary: column name -> list of values
        Dictionary<string, List<string>> ReadCSV(string data)
        {
            var lines = data.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
            var keys = lines[0].Split(',');
            var columns = keys.ToDictionary(k => k, k => new List<string>());

            foreach (var line in lines.Skip(1))
            {
                var values = line.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
                for (int i = 0; i < keys.Length; i++)
                {
                    columns[keys[i]].Add(values[i]);
                }
            }
            return columns;
        }

        // Transform the dictionary into a list of Person objects
        IEnumerable<Person> TransformData(Dictionary<string, List<string>> data)
        {
            var names = data["name"];
            var ages = data["age"].Select(int.Parse).ToList();

            var people = names.Zip(ages, (name, age) => new Person
            {
                Name = name.ToUpper(),
                Age = age,
                AgeGroup = age > 30 ? "old" : "young"
            }).ToList();

            return people;
        }

        var data = ReadCSV("name,age\nAlice,25\nBob,40");
        var people = TransformData(data);

        foreach (var person in people)
        {
            Console.WriteLine(person);
        }
    }
}